# e-Bay Kleinanzeigen Ad Data Cleaning Project

In this project, a 50.000 row data taken from Kaggle is to be worked on. The data dictionary is as follows

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which which year the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which which year the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim in this project is to clean and analyse used car listings.


In [2]:
import pandas as pd
import numpy as np

In [3]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

The data we have have 20 columns and almost all of the columns are in string from. On the other hand, there is no column which has more than 20% nulls.

In [6]:
autos.describe()

yearOfRegistration       powerPS  monthOfRegistration  nrOfPictures  \
count        50000.000000  50000.000000         50000.000000       50000.0   
mean          2005.073280    116.355920             5.723360           0.0   
std            105.712813    209.216627             3.711984           0.0   
min           1000.000000      0.000000             0.000000           0.0   
25%           1999.000000     70.000000             3.000000           0.0   
50%           2003.000000    105.000000             6.000000           0.0   
75%           2008.000000    150.000000             9.000000           0.0   
max           9999.000000  17700.000000            12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

First let's convert the column names from camelcase to snakecase

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
new_columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Seems all the column names are in snake case.

Now we need to further investigate the date to see if any attention is needed.

In [10]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Columns that have mostly one value are to be dropped like "seller" and "offer_type" columns which have 49999 of the same entry. 

In [11]:
autos.drop("seller", inplace = True, axis =1)
autos.drop("offer_type", inplace = True,axis =1)

In [12]:
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen'],
      dtype='object')

"seller" and "offer_type" columns have been removed. 

Now "price" and "odometer" columns include text data are to be converted.

In [13]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(float)

In [14]:
autos["price"].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

In [15]:
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype(float)
autos.rename({"odometer":"odometer_km"}, inplace=True, axis=1)

In [16]:
autos["odometer_km"].head()

0    150000.0
1    150000.0
2     70000.0
3     70000.0
4    150000.0
Name: odometer_km, dtype: float64

Looks everything is fine so far.
Now is the time to look for outliars at numerical columns, "odometer_km" and "price"

In [17]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [18]:
print("Number of unique price values are "+ str(autos["price"].nunique()))
print("Max and min price values are "+ str(autos["price"].max())+ " and " + str(autos["price"].min()))

Number of unique price values are 2357
Max and min price values are 99999999.0 and 0.0


It is controversy to have a free car, however a car with almost $100M is definitely an outlier. Let us further investigate

In [19]:
autos["price"].value_counts().head()

0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
Name: price, dtype: int64

In [20]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

Cars priced more than about $1M worth further examination. Now is the time to look for more data, which includes cars fit to this criterion. 

In [21]:
autos[autos["price"]>=999990][["name", "model", "price" ]]

name        model  \
514                     Ford_Focus_Turnier_1.6_16V_Style        focus   
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000       escort   
7814                                         Ferrari_F40          NaN   
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...          NaN   
22947                         Bmw_530d_zum_ausschlachten          NaN   
24384                            Schlachte_Golf_3_gt_tdi          NaN   
27371                                         Fiat_Punto        punto   
37585                                Volkswagen_Jetta_GT        jetta   
39377                        Tausche_volvo_v40_gegen_van          v40   
39705                        Tausch_gegen_gleichwertiges     s_klasse   
42221                                  Leasinguebernahme           c4   
43049                                      2_VW_Busse_T3  transporter   
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...       vectra   
47634                                        Ferrari_FXX          NaN   

            price  
514      999999.0  
2897   11111111.0  
7814    1300000.0  
11137  10000000.0  
22947   1234566.0  
24384  11111111.0  
27371  12345678.0  
37585    999990.0  
39377  12345678.0  
39705  99999999.0  
42221  27322222.0  
43049    999999.0  
47598  12345678.0  
47634   3890000.0

Now we need to make a web search if the prices are reasonable, at least for a couple of them like Ferrari's and Maserati. For the others, it is obvious that price information is not true and data should be cleaned by removing the rows.

_The web search showed that Ferrari cars at 7814 and 47634 have reasonable prices. Let us remove the rest._

In [22]:
autos[autos["price"]>=999990][["name", "model", "price" ]].index

Int64Index([  514,  2897,  7814, 11137, 22947, 24384, 27371, 37585, 39377,
            39705, 42221, 43049, 47598, 47634],
           dtype='int64')

In [23]:
# rows indexes other than 2 ferrari's have been listed below
rows_to_be_deleted = [  514,  2897, 11137, 22947, 24384, 27371, 37585, 39377,
            39705, 42221, 43049, 47598]

In [24]:
autos.drop(rows_to_be_deleted, axis = 0, inplace = True)

In [25]:
autos[autos["price"]>=999990][["name", "model", "price" ]]

name model      price
7814   Ferrari_F40   NaN  1300000.0
47634  Ferrari_FXX   NaN  3890000.0

12 data rows which are for the cars with problematic prices have been deleted. Now let us check odometer_km column. Now the remaining rows are 49988

In [26]:
autos["odometer_km"].describe()

count     49988.000000
mean     125732.575818
std       40042.406821
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [27]:
autos["odometer_km"].value_counts().head()

150000.0    32416
125000.0     5169
100000.0     2168
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64

In [28]:
autos["odometer_km"].value_counts().sort_index(ascending=True).head()

5000.0     967
10000.0    264
20000.0    784
30000.0    789
40000.0    818
Name: odometer_km, dtype: int64

Everything looks fine with "odometer_km" column. Let's move on to other columns.

In [29]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49988 entries, 0 to 49999
Data columns (total 18 columns):
date_crawled          49988 non-null object
name                  49988 non-null object
price                 49988 non-null float64
abtest                49988 non-null object
vehicle_type          44896 non-null object
registration_year     49988 non-null int64
gearbox               47310 non-null object
power_ps              49988 non-null int64
model                 47233 non-null object
odometer_km           49988 non-null float64
registration_month    49988 non-null int64
fuel_type             45509 non-null object
brand                 49988 non-null object
unrepaired_damage     40165 non-null object
ad_created            49988 non-null object
nr_of_pictures        49988 non-null int64
postal_code           49988 non-null int64
last_seen             49988 non-null object
dtypes: float64(2), int64(5), object(11)
memory usage: 7.2+ MB


Here it is seen that there are some columns with null values, which are *vehicle_type*, *gearbox*, *model*, *fuel_type* and *unrepaired_damage*. 

Now is the time to deal with columns including dates. There are five of them which are *date_crawled*, *ad_created*, *last_seen*, *registration_month* and *registration_year*

The last two are already in numeric form. However the first three are in string. So let us explore those three first.

In [30]:
autos[["date_crawled","ad_created", "last_seen"] ].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Data for those three columns are sortable. Since only day matters, we can extract the first ten chars.

In [31]:
#first ten characters are put in new columns
autos["date_crawled_day"]=autos["date_crawled"].str[:10]
autos["ad_created_day"]=autos["ad_created"].str[:10]
autos["last_seen_day"]=autos["last_seen"].str[:10]

In [32]:
# now is the time to explore *date_crawled_day* column
autos["date_crawled_day"].value_counts(normalize = True, dropna=False)

2016-04-03    0.038689
2016-03-20    0.037829
2016-03-21    0.037489
2016-03-12    0.036769
2016-03-14    0.036629
2016-04-04    0.036529
2016-03-07    0.035969
2016-04-02    0.035408
2016-03-19    0.034908
2016-03-28    0.034848
2016-03-29    0.034148
2016-03-15    0.033988
2016-04-01    0.033808
2016-03-30    0.033628
2016-03-08    0.033268
2016-03-09    0.033208
2016-03-22    0.032908
2016-03-11    0.032488
2016-03-26    0.032488
2016-03-23    0.032388
2016-03-10    0.032128
2016-03-31    0.031908
2016-03-25    0.031748
2016-03-17    0.031508
2016-03-27    0.031047
2016-03-16    0.029507
2016-03-24    0.029107
2016-03-05    0.025386
2016-03-13    0.015564
2016-03-06    0.013943
2016-04-05    0.013103
2016-03-18    0.013063
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled_day, dtype: float64

In [33]:
autos["date_crawled_day"].describe()

count          49988
unique            34
top       2016-04-03
freq            1934
Name: date_crawled_day, dtype: object

In [34]:
# now is the time to explore *ad_created_day* column
autos["ad_created_day"].value_counts(normalize = True, dropna=False)

2016-04-03    0.038929
2016-03-20    0.037869
2016-03-21    0.037689
2016-04-04    0.036889
2016-03-12    0.036609
2016-03-14    0.035228
2016-04-02    0.035088
2016-03-28    0.034968
2016-03-07    0.034748
2016-03-29    0.034108
2016-03-19    0.033848
2016-04-01    0.033808
2016-03-15    0.033748
2016-03-30    0.033448
2016-03-08    0.033308
2016-03-09    0.033228
2016-03-11    0.032788
2016-03-22    0.032768
2016-03-26    0.032568
2016-03-23    0.032188
2016-03-31    0.031908
2016-03-25    0.031888
2016-03-10    0.031868
2016-03-17    0.031187
2016-03-27    0.030907
2016-03-16    0.030007
2016-03-24    0.029087
2016-03-05    0.023046
2016-03-13    0.016924
2016-03-06    0.015124
                ...   
2016-02-25    0.000060
2016-02-02    0.000040
2016-02-24    0.000040
2016-02-18    0.000040
2016-02-14    0.000040
2016-02-09    0.000040
2016-02-05    0.000040
2016-01-10    0.000040
2016-02-20    0.000040
2016-02-26    0.000040
2015-11-10    0.000020
2016-02-08    0.000020
2016-02-22 

In [35]:
autos["ad_created_day"].describe()

count          49988
unique            76
top       2016-04-03
freq            1946
Name: ad_created_day, dtype: object

In [36]:
# now is the time to explore *last_seen* column
autos["last_seen_day"].value_counts(normalize = True, dropna=False)

2016-04-06    0.220973
2016-04-07    0.130951
2016-04-05    0.124310
2016-03-17    0.027927
2016-04-03    0.025366
2016-04-02    0.024886
2016-03-30    0.024846
2016-04-04    0.024626
2016-03-31    0.023826
2016-03-12    0.023806
2016-04-01    0.023106
2016-03-29    0.022325
2016-03-22    0.021585
2016-03-28    0.020845
2016-03-21    0.020725
2016-03-20    0.020705
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-23    0.018584
2016-03-26    0.016964
2016-03-16    0.016444
2016-03-27    0.016024
2016-03-15    0.015884
2016-03-19    0.015744
2016-03-14    0.012803
2016-03-11    0.012523
2016-03-10    0.010763
2016-03-09    0.009842
2016-03-13    0.008982
2016-03-08    0.007582
2016-03-18    0.007422
2016-03-07    0.005361
2016-03-06    0.004421
2016-03-05    0.001080
Name: last_seen_day, dtype: float64

In [37]:
autos["last_seen_day"].describe()

count          49988
unique            34
top       2016-04-06
freq           11046
Name: last_seen_day, dtype: object

Everything looks fine with these three date columns. Now let us investigate *registration_year* data

In [38]:
autos["registration_year"].describe()

count    49988.000000
mean      2005.075478
std        105.725062
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

It is observed that min and max values are problematic. Let us first sort and look at the head and tail of the sorted list.

In [44]:
autos["registration_year"].sort_values(ascending = False).head(30)

33950    9999
8012     9999
14341    9999
38076    9999
6308     9996
49910    9000
13559    9000
25003    8888
8360     6200
27618    5911
24519    5000
22799    5000
4164     5000
49153    5000
42079    4800
453      4500
4549     4100
27578    2800
5763     2019
49185    2019
38342    2019
28120    2018
40375    2018
14264    2018
7272     2018
31281    2018
36227    2018
28088    2018
8450     2018
36258    2018
Name: registration_year, dtype: int64

In [45]:
autos["registration_year"].sort_values(ascending = True).head(30)

22316    1000
49283    1001
24511    1111
35238    1500
10556    1800
32585    1800
45157    1910
3679     1910
46213    1910
42181    1910
33295    1910
28693    1910
30781    1910
22659    1910
15898    1910
21416    1927
22101    1929
11246    1931
2221     1934
2573     1934
26607    1937
39725    1937
21421    1937
23804    1937
26103    1938
24855    1939
13963    1941
25792    1941
11585    1943
11047    1948
Name: registration_year, dtype: int64

Since the cars cannot be first registered after the listing was seen and *last_seen* data gives that all the cars in the list are last seen on 2016, there cannot be any data after 2016 in this list. With regards to the beginning, for antiuqe cars, it is possible to be registered on 1910. Therefore we will remove the rest of the data which is outside this interval.

In [58]:
#rows that are out of range are determined
out_of_range_autos = autos[(autos["registration_year"]<1900) | 
                           (autos["registration_year"]>2016)]
out_of_range_autos.shape
# 1970 rows are to be deleted

(1970, 21)

In [62]:
# out of range autos are dropped here
autos.drop(out_of_range_autos.index, inplace = True)

In [63]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km      ...       fuel_type       brand unrepaired_damage  \
0  andere     150000.0      ...             lpg     peugeot              nein   
1     7er     150000.0      ...          benzin         bmw              nein   
2    golf      70000.0      ...          benzin  volkswagen              nein   
3  fortwo      70000.0      ...          benzin       smart              nein   
4   focus     150000.0      ...          benzin        ford              nein   

            ad_created nr_of_pictures  postal_code            last_seen  \
0  2016-03-26 00:00:00              0        79588  2016-04-06 06:45:54   
1  2016-04-04 00:00:00              0        71034  2016-04-06 14:45:08   
2  2016-03-26 00:00:00              0        35394  2016-04-06 20:15:37   
3  2016-03-12 00:00:00              0        33729  2016-03-15 03:16:28   
4  2016-04-01 00:00:00              0        39218  2016-04-01 14:38:50   

  date_crawled_day ad_created_day last_seen_day  
0       2016-03-26     2016-03-26    2016-04-06  
1       2016-04-04     2016-04-04    2016-04-06  
2       2016-03-26     2016-03-26    2016-04-06  
3       2016-03-12     2016-03-12    2016-03-15  
4       2016-04-01     2016-04-01    2016-04-01  

[5 rows x 21 columns]

In [64]:
autos["registration_year"].value_counts(normalize=True)

2000    0.069849
2005    0.062789
1999    0.062435
2004    0.056999
2003    0.056791
2006    0.056396
2001    0.056271
2002    0.052751
1998    0.051085
2007    0.047982
2008    0.046462
2009    0.043671
1997    0.042234
2011    0.034029
2010    0.033258
1996    0.030072
2012    0.027552
2016    0.027406
1995    0.027323
2013    0.016785
2014    0.013849
1994    0.013745
1993    0.009267
2015    0.008309
1990    0.008226
1992    0.008143
1991    0.007414
1989    0.003769
1988    0.002957
1985    0.002166
          ...   
1966    0.000458
1977    0.000458
1975    0.000396
1969    0.000396
1965    0.000354
1964    0.000250
1910    0.000187
1963    0.000187
1959    0.000146
1961    0.000125
1956    0.000104
1958    0.000083
1937    0.000083
1962    0.000083
1950    0.000062
1954    0.000042
1941    0.000042
1934    0.000042
1957    0.000042
1951    0.000042
1955    0.000042
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.0000

Now the *registration year* column look more reasonable. Comparably new cars show up more than old ones.

Now is the time to work on *brand* column data. Let us have a look at the unique values at that column. Then let us form a dictionary which holds a mean price for the selected brands.

In [69]:
print("There are " +str(autos["brand"].nunique()) +" different car brands in this list")

There are 40different car brands in this list


In [81]:
#Now let us examine the frequency of appereances of those brands

autos["brand"].value_counts(normalize = True).head(10)
# this list is already sorted and let us choose those which has more than 3%

volkswagen       0.212108
bmw              0.110021
opel             0.108168
mercedes_benz    0.095360
audi             0.086405
ford             0.069766
renault          0.047357
peugeot          0.029531
fiat             0.025865
seat             0.018181
Name: brand, dtype: float64

In [79]:
#brands of which mean price will be calculated are as follows
selected_brands = ["volkswagen", "bmw", "opel", "mercedes_benz",
                   "audi", "ford", "renault"]

In [89]:
# mean prices for selected brands have been recorded in the dictionary
mean_prices = {}

for item in selected_brands:
    mean_prices[item] = (autos[autos["brand"]==item]["price"].sum()) / (autos["brand"].value_counts()[item])
    
    
mean_prices

{'audi': 9093.65003615329,
 'bmw': 8102.536248343744,
 'ford': 3652.095223880597,
 'mercedes_benz': 8485.239571958942,
 'opel': 2876.716403542549,
 'renault': 2395.4164467897976,
 'volkswagen': 5231.081983308787}

Average prices are given in the dictionary above. The results seem reasonable. Mercedes Benz, Audi and BMW are expensive, Volkswagen is a little cheaper, while Ford, Opel and Renault are more cheaper.
Now let us examine average milage of the cars to see if there is a link between prices and milages



In [93]:
# First let us do the aggregation on milage. Pretty much the same as we did for prices
# mean milages for selected brands have been recorded in the dictionary
mean_milages = {}

for item in selected_brands:
    mean_milages[item] = (autos[autos["brand"]==item]["odometer_km"].sum()) / (autos["brand"].value_counts()[item])
    
    
mean_milages

{'audi': 129287.78018799711,
 'bmw': 132431.38368351315,
 'ford': 124068.65671641791,
 'mercedes_benz': 130856.0821139987,
 'opel': 129223.14208702349,
 'renault': 128183.81706244503,
 'volkswagen': 128724.10407461954}

In [102]:
# Now lets merge both dictionaries into a data frame to do the analysis

mean_prices_series = pd.Series(mean_prices)
mean_milages_series = pd.Series(mean_milages)

means = pd.DataFrame(mean_prices_series, columns=["mean_price"])
means["mean_milages"]=mean_milages_series

means.astype(int)

mean_price  mean_milages
audi                 9093        129287
bmw                  8102        132431
ford                 3652        124068
mercedes_benz        8485        130856
opel                 2876        129223
renault              2395        128183
volkswagen           5231        128724

Although mean prices change depending on the brand but milages are all around 120K-135K. Therefore there is no obvious link between mean milage of the cars and its prices. 